chore: Added tests for GetAdminUrlForLanguage template tag

filipweidemann opened the pull request:
## Description
This PR catches up on #8044 and adds two new tests targeting the regressed template tag `GetAdminUrlForLanguage`.
I think this is worth it. Please let me know what you think.

## Related resources
#8044 

## Checklist
* [x] I have opened this pull request against ``develop-4``
* [x] I have added or modified the tests when changing logic
* [x] I have followed [the conventional commits guidelines](https://www.conventionalcommits.org/) to add meaningful information into the changelog
* [x] I have read the [contribution guidelines ](https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst) and I have joined the channel #pr-reviews on our [Discord Server](https://discord-pr-review-channel.django-cms.org) to find a “pr review buddy” who is going to review my pull request.

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
+        with force_language('de'):
+            url = GetAdminUrlForLanguage.get_value(None, context={'request': request}, page=page, language="de")
+            # We expect that getting the German URL will return the German page content (change link, not a create link)
+            self.assertEqual(url, '/de/admin/cms/pagecontent/2/change/')
```
```suggestion
            self.assertEqual(url, f'/de/admin/cms/pagecontent/{german_content.pk}/change/')
```
It's unsafe to make assumptions on pk's in testing environments.

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
+        with force_language('en'):
+            url = GetAdminUrlForLanguage.get_value(None, context={'request': request}, page=page, language="en")
+            # Same goes for the English URL
+            self.assertEqual(url, '/en/admin/cms/pagecontent/1/change/')
```
```suggestion
            self.assertEqual(url, f'/en/admin/cms/pagecontent/{english_content.pk}/change/')
```

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
+        # This creates a new page and the corresponding PageContent object for lang 'en', both with pk=1
+        page = create_page("AdminURLTestPage English Content", "nav_playground.html", "en")
+
+        # German PageContent will have pk=2
```
```suggestion
```
It will not necessarily.

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
 class TemplatetagTests(CMSTestCase):
+    def test_get_admin_url_for_language(self):
+        # This creates a new page and the corresponding PageContent object for lang 'en', both with pk=1
+        page = create_page("AdminURLTestPage English Content", "nav_playground.html", "en")
```
```suggestion
        page = create_page("AdminURLTestPage English Content", "nav_playground.html", "en")
        
        english_content = page.pagecontent_set(manager="admin_manager").first()
```

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
+
+    def test_admin_pagecontent_language_tab_urls(self):
+        # Same setup as in the templatetag test above
+        page = create_page('Test', 'nav_playground.html', 'en')
```
```suggestion
        page = create_page('Test', 'nav_playground.html', 'en')
        english_coontent = page.pagecontent_set(manager="admin_manager").first()
```

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
+            {% get_admin_url_for_language page_obj 'fr' %}
+        """
+        output = self.render_template_obj(template, {'page_obj': page}, request)
+        self.assertIn('/en/admin/cms/pagecontent/1/change/', output)
```
```suggestion
        self.assertIn(f'/en/admin/cms/pagecontent/{english_content.pk}/change/', output)
```

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,48 @@
+        """
+        output = self.render_template_obj(template, {'page_obj': page}, request)
+        self.assertIn('/en/admin/cms/pagecontent/1/change/', output)
+        self.assertIn('/en/admin/cms/pagecontent/2/change/', output)
```
```suggestion
        self.assertIn(f'/en/admin/cms/pagecontent/{german_content.pk}/change/', output)
```

fsbraun reviewed (commented):
Yes, this is a good idea! 

However, you'll need to avoid assumptions on the primary keys.

filipweidemann commented:
@fsbraun Yup, that's right. I added your suggestions in one commit & fixed a spelling issue. Should be fine now! Do you have any other suggestions?

fsbraun commented:
@filipweidemann What exactly is the difference between the two tests? Do we need them both?

I can imagine two tests (based on either of yours).
1. In the test setup call `page.get_content_obj("en")` before calling the template tag.
2. In the test setup call `page.get_admin_content("en")` before calling the template tag.

With a partially filled cache by the two calls, the tests would fail.

filipweidemann commented:
@fsbraun Yeah wow, I actually forgot the third case, you're right. Was thinking about this before implementing it. I'll add it as well!

About the current ones: I can probably remove the first one and only stick with the "faked" template one. My idea was that we also check if it can be forced to craft the links according to our language but let's be real, it actually is not doing anything here because it uses Django's `reverse` function. So what I am actually testing is Django and yeah, this is kind of unnecessary.

I'll remove the first one and replace it with the one you suggested because this one is actually missing right now!

filipweidemann commented:
@fsbraun okay so I deleted the first one and added the method calls before rendering the template.

I also verified that this catches the regression by temporarily adding `del self.admin_content_cache["de"]` inside the `set_admin_content_cache`. The test immediately failed.

This is cool. Anything else?

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,28 @@
+        german_content = create_page_content("de", "AdminURLTestPage German Content", page)
+
+        # Try to fill the cache with partial data (this should not be possible)
+        page.get_content_obj(language='en')
```
```suggestion
        page.get_content_obj(language='en')  # should not affect admin template tag
```

fsbraun commented on cms/tests/test_templatetags.py:
```diff
@@ -43,6 +43,28 @@
+
+        # Try to fill the cache with partial data (this should not be possible)
+        page.get_content_obj(language='en')
+        page.get_admin_content(language='en')
```
```suggestion
        page.get_admin_content(language='en')  # Should fill the whole cache
```

fsbraun reviewed (approved):
Nice! Looks good to me!! Thank you so much!
